### PR TITLE
Prevent acting on non existing tables (Fix #2663)

### LINKED ISSUE #2663: Symphony wants to manipulate tables that do not exist
In `2.7.x` @michael-e has introduced »table locking« to prevent data corruption. This apparently leads to problems with fields that do not create an entry table since they do not store any data. 

I’ve encountered an error with the [system_date_fields](https://github.com/symphonists/system_date_fields/issues/20)-extension. See link for the corresponding issue.

To quote Michael:

> In your setup MySQL vomits on the fact that Symphony wants to manipulate tables which do not exist. Symphony has always assumed that these tables exist.

What would be the preferable solution? 

1. symphony checking if the table actually exists, or 
2. `system_date_fields` creating an empty table?

### PR DESCRIPTION
**Pull request by nitriques**
This change makes sure that the changes introduced in 4b5067f38 do not
cause a bug when the table does not exists. The old code (pre 2.7.0)
discarded all errors on delete, then checked if it had data to be
inserted. If not, the insert statement would be skipped.

The locking mechanism did add a lock statement before the delete, and
added logging all exceptions. That resulted in log "vomit".

This changes propose 2 ways of fixing this log problem:
1. It moves the empty data checks BEFORE the lock (so before the delete
as well).
2. It checks in the database if the table exist before doing anything.

4b5067f38 also un-lock the table before doing the insert. I think that
it should still maintain the lock until both delete and insert are done.

Fixes #2663

cc @animaux @michael-e

### CODE REVIEW

**Comment by animaux** [comment]
No errors, entry saves fine. Works at least in conjunction with `system_date_fields`.

**Comment by jonmifsud** [comment]
Would adding a `containsEntryTable` function requirement on Field do the
trick in a more efficient manner?

We can add the default function definition which checks for a table
existence and stores it within the field object (if that is re-used) or
else default to true and only fields which don't have this would override
this parameter.





*Jonathan Mifsud*
Business digitization expert
  T +356 9983 0394 <+35699830394> E jon@maze.digital W maze.digital

File: symphony/lib/toolkit/class.mysql.php [hunk 1]
```diff
@@ -882,8 +883,9 @@ public function fetchVar($column, $offset = 0, $query = null)
      */
     public function tableContainsField($table, $field)
     {
+        $table = MySQL::cleanValue($table);
+        $field = MySQL::cleanValue($field);
```
Comment (michael-e) [inline symphony/lib/toolkit/class.mysql.php]: Ideally, this should be a separate PR, because it has nothing to do with the topic.

**Comment by michael-e** [comment]
@jonmifsud: Implementing a `containsEntryTable` method for fields would change a lot. Extensions would need to be updated. However, I see the advantage—it wouldn't slow things down.

So is @nitriques solution slowing things down too much? I wouldn't think so either. Having 3 queries per field and entry instead of 2 (delete/insert) may sound a lot. But I would expect the additional query `SHOW TABLES LIKE…` to be rather fast compared to deleting or especially inserting.

**Comment by jonmifsud** [comment]
@michael-e would it change that much? As in if you implement the default
behaviour to be true (as it was before) most would still work out of the
box right? I think only this field has no `table` and maybe a handful of
others. Obviously we can always go with a `phase1` which is backwards
compatible then in the next major version add the function.





*Jonathan Mifsud*
Business digitization expert
  T +356 9983 0394 <+35699830394> E jon@maze.digital W maze.digital

**Comment by michael-e** [comment]
@jonmifsud: I agree—most field extensions do have data tables, of course. But I have the feeling that @nitriques wants a more "universal" solution like "Never act on tables which do not exist. Never.". Let's see what he thinks.

File: symphony/lib/toolkit/class.mysql.php (hunk 1 above)
Comment (nitriques) [inline symphony/lib/toolkit/class.mysql.php]: Yeah, guilty, I should have `git add . -p`

**Comment by nitriques** [comment]
@brendo XML importer does update stuff ? (I've only used it for inserting)

> @nitriques wants a more "universal" solution like "Never act on tables that do not exist. Never.". Let's see what he thinks.

I sure do. Having those kind of bugs in production environment can cost a lot.

I like @jonmifsud idea tho. I could speed things up. Let's add it (?)

**Comment by nitriques** [comment]
> BTW, my commit 4b5067f did not un-lock the table before doing the insert

@michael-e yup sorry interpreted the code wrong. But right now, no lock occurs for empty field (thus reducing the number of queries!)

**Comment by michael-e** [comment]
> Let's add it (?)

Oh, I have nothing against it, on the contrary! If you can do the work…

**Comment by nitriques** [comment]
I am all for it, I'll do it. Great!

**Comment by nitriques** [comment]
We already have `requiresSQLGrouping`. What about `requiresEntryTable` ?

**Comment by michael-e** [comment]
The naming scheme for the tables is `tbl_entries_data_[:id]`, so what about `requiresEntriesDataTable` or simply `hasEntriesDataTable`?

**Comment by nitriques** [comment]
I'd vote for `requiresEntriesDataTable`, but the current 'table' related method is named `createTable` and we also have a `tearDown` method which, strangely, does not drop the table it created.... few naming really is the second hardest problem there is.

But our best bet would be to stick with what's in there right now and name it `requiresTable`

**Comment by nitriques** [comment]
@michael-e another naming problem here :(

**Comment by michael-e** [comment]
Yes, naming here can be either consistent (at least a bit) or clear… I promise that I won't cry, whatver you decide.

**Comment by nitriques** [comment]
> I promise that I won't cry, whatver you decide.

Thanks hahah

**Comment by nitriques** [comment]
I'll send a separate PR for the new function.

**Comment by nitriques** [comment]
See #2678 for a follow up. cc @jonmifsud

**Comment by pointybeard** [comment]
A bit late to the party, but this PR appears to break the ability to empty an optional field of its value. https://github.com/symphonycms/symphony-2/commit/257aab2deb688c4b6c68805b102dabd4ba880c7e#commitcomment-23450932
